File: rust/src/order.rs
use std::cell::RefCell;
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::rc::Rc;

#[derive(Clone, Copy, Eq, PartialEq, Debug)]
pub enum Side {
    Buy = 1,
    Sell = -1
}

impl Side {
    pub fn as_f64(&self) -> f64 {
        match self {
            Side::Buy => 1f64,
            Side::Sell => -1f64
        }
    }

    pub fn as_f32(&self) -> f32 {
        match self {
            Side::Buy => 1f32,
            Side::Sell => -1f32
        }
    }
}

#[derive(Clone, Copy, Eq, PartialEq, Debug)]
pub enum Status {
    None = 0,
    New = 1,
    Expired = 2,
    Filled = 3,
    Canceled = 4,
    PartiallyFilled = 5
}

#[derive(Clone, Copy, Eq, PartialEq, Debug)]
pub enum TimeInForce {
    GTC = 0,
    GTX = 1,
    FOK = 2,
    IOC = 3,
}

#[derive(Clone, Copy, Eq, PartialEq, Debug)]
pub enum OrdType {
    Limit = 0,
    Market = 1,
}

#[derive(Clone, Debug)]
pub struct Order<Q> where Q: Sized + Clone {
    pub qty: f32,
    pub leaves_qty: f32,
    pub price_tick: i32,
    pub tick_size: f32,
    pub side: Side,
    pub time_in_force: TimeInForce,
    pub exch_timestamp: i64,
    pub status: Status,
    pub local_timestamp: i64,
    pub req: Status,
    pub exec_price_tick: i32,
    pub exec_qty: f32,
    pub order_id: i64,
    pub q: Q,
    pub maker: bool,
    pub order_type: OrdType
}

impl<Q> Order<Q> where Q: Clone + Default {
    pub fn new(
        order_id: i64,
        price_tick: i32,
        tick_size: f32,
        qty: f32,
        side: Side,
        order_type: OrdType,
        time_in_force: TimeInForce,
    ) -> Self {
        Self {
            qty,
            leaves_qty: qty,
            price_tick,
            tick_size,
            side,
            time_in_force,
            exch_timestamp: 0,
            status: Status::None,
            local_timestamp: 0,
            req: Status::None,
            exec_price_tick: 0,
            exec_qty: 0.0,
            order_id,
            q: Q::default(),
            maker: false,
            order_type,
        }
    }

    pub fn price(&self) -> f32 {
        self.price_tick as f32 * self.tick_size
    }

    pub fn exec_price(&self) -> f32 {
        self.exec_price_tick as f32 * self.tick_size
    }

    pub fn cancellable(&self) -> bool {
        self.status == Status::New && self.req == Status::None
    }
}

#[derive(Clone, Debug)]
pub struct OrderBus<Q> where Q: Clone {
    order_list: Rc<RefCell<Vec<(Order<Q>, i64)>>>,
    orders: Rc<RefCell<HashMap<i64, i64>>>,
}

impl<Q> OrderBus<Q> where Q: Clone {
    pub fn new() -> Self {
        Self {
            order_list: Default::default(),
            orders: Default::default(),
        }
    }

    pub fn frontmost_timestamp(&self) -> i64 {
        self.order_list.borrow().get(0).map(|(_order, ts)| *ts).unwrap_or(i64::MAX)
    }

    pub fn append(&mut self, order: Order<Q>, timestamp: i64) {
        *self.orders.borrow_mut().entry(order.order_id).or_insert(0) += 1;
        self.order_list.borrow_mut().push((order, timestamp));
    }

    pub fn get_head_timestamp(&self) -> Option<i64> {
        if let Some((_order, recv_ts)) = self.order_list.borrow().get(0) {
            Some(*recv_ts)
        } else {
            None
        }
    }

    pub fn get(&self, order_id: i64) -> Option<i64> {
        for (order, recv_ts) in self.order_list.borrow().iter() {
            if order.order_id == order_id {
                return Some(*recv_ts);
            }
        }
        None
    }

    pub fn reset(&mut self) {
        self.order_list.borrow_mut().clear();
        self.orders.borrow_mut().clear();
    }

    pub fn len(&self) -> usize {
        self.order_list.borrow().len()
    }

    pub fn remove(&mut self, index: usize) -> Order<Q> {
        let (order, _) = self.order_list.borrow_mut().remove(index);
        if let Entry::Occupied(mut entry) = self.orders.borrow_mut().entry(order.order_id) {
            let value = entry.get_mut();
            *value -= 1;
            if *value <= 0 {
                entry.remove();
            }
        }
        order
    }

    pub fn contains_key(&self, order_id: i64) -> bool {
        self.orders.borrow().contains_key(&order_id)
    }
}

File: hftbacktest/order.py
import numpy as np
from numba import float32, int32, int64, uint8, from_dtype
from numba.experimental import jitclass
from numpy.typing import NDArray

UNSUPPORTED = 255

BUY = 1
SELL = -1

NONE = 0
NEW = 1
EXPIRED = 2
FILLED = 3
CANCELED = 4
PARTIALLY_FILLED = 5
MODIFY = 6
REJECTED = 7

GTC = 0  # Good 'till cancel
GTX = 1  # Post only
FOK = 2  # Fill or kill
IOC = 3  # Immediate or cancel

LIMIT = 0
MARKET = 1

order_dtype = np.dtype([
    ('qty', 'f4'),
    ('leaves_qty', 'f4'),
    ('price_tick', 'i4'),
    ('tick_size', 'f4'),
    ('exch_timestamp', 'i8'),
    ('local_timestamp', 'i8'),
    ('exec_price_tick', 'i4'),
    ('exec_qty', 'f4'),
    ('order_id', 'i8'),
    ('_q1', 'u8'),
    ('_q2', 'u8'),
    ('maker', 'bool'),
    ('order_type', 'u1'),
    ('req', 'u1'),
    ('status', 'u1'),
    ('side', 'u1'),
    ('time_in_force', 'u1'),
])


@jitclass
class Order:
    _arr: from_dtype(order_dtype)[:]

    def __init__(self, arr: order_dtype):
        self._arr = arr

    @property
    def price(self) -> float32:
        return self._arr[0].price_tick * self._arr[0].tick_size

    @property
    def exec_price(self) -> float32:
        return self._arr[0].exec_price_tick * self._arr[0].tick_size

    @property
    def cancellable(self) -> bool:
        return (self._arr[0].status == NEW or self._arr[0].status == PARTIALLY_FILLED) and self._arr[0].req == NONE

    @property
    def qty(self) -> float32:
        return self._arr[0].qty

    @property
    def leaves_qty(self) -> float32:
        return self._arr[0].leaves_qty

    @property
    def price_tick(self) -> int32:
        return self._arr[0].price_tick

    @property
    def tick_size(self) -> float32:
        return self._arr[0].price_tick

    @property
    def exch_timestamp(self) -> int64:
        return self._arr[0].exch_timestamp

    @property
    def local_timestamp(self) -> int64:
        return self._arr[0].local_timestamp

    @property
    def exec_price_tick(self) -> int32:
        return self._arr[0].exec_price_tick

    @property
    def exec_qty(self) -> float32:
        return self._arr[0].exec_qty

    @property
    def order_id(self) -> int64:
        return self._arr[0].order_id

    @property
    def order_type(self) -> uint8:
        return self._arr[0].order_type

    @property
    def req(self) -> uint8:
        return self._arr[0].req

    @property
    def status(self) -> uint8:
        return self._arr[0].status

    @property
    def side(self) -> uint8:
        return self._arr[0].side

    @property
    def time_in_force(self) -> uint8:
        return self._arr[0].time_in_force
